server/hotrod/src/main/java/org/infinispan/server/hotrod/TransactionRequestProcessor.java

```diff
@@ -38,41 +36,70 @@
       super(channel, executor, server);
    }
 
+   private void writeTransactionResponse(HotRodHeader header, int value) {
+      writeResponse(header, createTransactionResponse(header, value));
+   }
+
    /**
     * Handles a rollback request from a client.
-    * @param header
-    * @param subject
-    * @param xid
     */
    void rollbackTransaction(HotRodHeader header, Subject subject, XidImpl xid) {
-      AdvancedCache<byte[], byte[]> cache = server.cache(header, subject);
-      validateConfiguration(cache);
-      executor.execute(() -> rollbackTransactionInternal(header, cache, xid));
+      RollbackTransactionOperation operation = new RollbackTransactionOperation(header, server, subject, xid,
+            this::writeTransactionResponse);
+      executor.execute(operation);
    }
 
-   private void rollbackTransactionInternal(HotRodHeader header, AdvancedCache<byte[], byte[]> cache, XidImpl xid) {
-      try {
-         writeResponse(header, finishTransaction(header, new RollbackTransactionDecodeContext(cache, xid)));
-      } catch (Throwable t) {
-         writeException(header, t);
-      }
+   /**
+    * Handles a commit request from a client
+    */
+   void commitTransaction(HotRodHeader header, Subject subject, XidImpl xid) {
+      CommitTransactionOperation operation = new CommitTransactionOperation(header, server, subject, xid,
+            this::writeTransactionResponse);
+      executor.execute(operation);
    }
 
    /**
     * Handles a prepare request from a client
-    * @param header
-    * @param subject
-    * @param xid
-    * @param onePhaseCommit
-    * @param writes
     */
-   void prepareTransaction(HotRodHeader header, Subject subject, XidImpl xid, boolean onePhaseCommit, List<TransactionWrite> writes) {
+   void prepareTransaction(HotRodHeader header, Subject subject, XidImpl xid, boolean onePhaseCommit,
+         List<TransactionWrite> writes, boolean recoverable, long timeout) {
       AdvancedCache<byte[], byte[]> cache = server.cache(header, subject);
       validateConfiguration(cache);
-      executor.execute(() -> prepareTransactionInternal(header, cache, xid, onePhaseCommit, writes));
+      executor.execute(() -> prepareTransactionInternal(header, cache, xid, onePhaseCommit, writes, recoverable, timeout));
+   }
+
+   void forgetTransaction(HotRodHeader header, Subject subject, XidImpl xid) {
+      //TODO authentication?
+      GlobalTxTable txTable = server.getCacheManager().getGlobalComponentRegistry().getComponent(GlobalTxTable.class);
+      executor.execute(() -> {
+         try {
+            txTable.forgetTransaction(xid);
```

Comment:
I really wish our internal tx stuff was async... OTOH the first prototype of project loom is already here, maybe we won't need to make it async after all :)

http://mail.openjdk.java.net/pipermail/loom-dev/2018-July/000061.html

Reply:
The first LTS JDK with loom will be 14 (if everything goes well), that's 2 years...